### DISCUSSION TITLE
Prefs change for the ChromeOptions not working

### DISCUSSION THREAD

**Discussion by yzvezdin** [webdriverio/webdriverio#7347] [Technical Discussion]
Hi guys,
We need to change the download path and set it via prefs as all the documentation says. However these values are not implemented in Chrome. Default path to Downloads is set and not changed
```
'goog:chromeOptions': {
            args: ['--disable-extensions', '--auth-server-whitelist="_"', '-window-position=0,0', '--start-maximized','--window-size=1920,1080'].concat(additionalArgs),
            prefs: { 'download.default_directory': `${process.cwd()}\\downloads` }
        }
```

**Comment by christian-bromann** [comment]
@yzvezdin the following worked for me:

```js
const { remote } = require('webdriverio')

let browser

const opts = {
    // automationProtocol: 'devtools',
    capabilities: {
        browserName: 'chrome',
        'goog:chromeOptions': {
            prefs: {
                'download.default_directory': __dirname
            }
        }
    }
}

;(async function () {
    browser = await remote(opts)
    await browser.url('https://the-internet.herokuapp.com/download')
    const link = await browser.$("a[href*='some-file.txt']")
    await link.click()
    await browser.pause(5000)

    await browser.deleteSession()
})().catch(async (e) => {
    console.error(e.stack)
    await browser.deleteSession()
})
```

Note: this of course only works for local test. If you run tests in a remote location (e.g. Selenium Grid or a vendor like Sauce Labs) this obviously won't work.

**Reply by yzvezdin** [reply]
Thank you. Will doublecheck.

**Reply by yzvezdin** [reply]
![image](https://user-images.githubusercontent.com/16023406/131103956-60cb262d-cb11-41db-8b78-5ea1e44c9494.png)
So I have used your code and... it still saved to the Downloads folder of the system :(

**Reply by christian-bromann** [reply]
> it still saved to the Downloads folder of the system :(

Can you try with `__dirname` to see if something with the path is wrong?

PS: don't share screenshots as it makes it difficult to copy your code into someones IDE. it is better to use the available markdown features or share a gist.

**Reply by yzvezdin** [reply]
Didn't work either. However path contains '-' and spaces. But it wasn't an issue before.

**Reply by christian-bromann** [reply]
Not sure then, could be a Windows specific problem. There is nothing WebdriverIO can do here.

**Reply by yzvezdin** [reply]
But the problem is - I check Chrome download path in these instances and they are default :( So parameters are not provided for Chrome. For Protractor these were passed properly.

**Reply by christian-bromann** [reply]
Well, then you must use WebdriverIO without Chromedriver. Indeed when using WebdriverIO with Puppeteer setting prefs is not supported. I created an issue: https://github.com/webdriverio/webdriverio/issues/7351, please get involved to get this fixed. As a workaround you can just run your tests with Chromedriver.

**Reply by yzvezdin** [reply]
Thank you, sir! :)

**Reply by NavyMint** [reply]
Hi, @christian-bromann .  Thank you for fix in chrome-launcher. I've checked its already merged and released in version 0.14.1. But when I'm testing download its still not working for webdriverIO with Jasmine.
**closed issue**: https://github.com/webdriverio/webdriverio/issues/7351
**my configuration**: https://github.com/NavyMint/reproduce-screenshot-error/blob/main/wdio.conf.ts
**test based on your example posted above**: https://github.com/NavyMint/reproduce-screenshot-error/blob/main/test/specs/download.e2e.ts

**Reply by Yurii397** [reply]
Hey guys, 
I'm upgrading the Chrome version from 95 to 112 and am faced with strange behavior, that in v112 I can't set a new download directory using the following options :
     prefs: {
          directory_upgrade: true,
          prompt_for_download: false,
          'download.default_directory': global.downloadFolder,
        }, 
The most strange thing is that it only doesn't work in headless mode. In headed it works well. 
Tried both Chrome headless (new and old one) - https://developer.chrome.com/articles/new-headless/

BTW, I'm using standalone-service without Pupetter

Is there any idea what was changed or how to handle it?
cc. @christian-bromann  @NavyMint

**Reply by christian-bromann** [reply]
no

**Reply by Yurii397** [reply]
> no

is it a direct Chrome issue?
and maybe any suggestion where it can be reported plz?

**Reply by christian-bromann** [reply]
you might want to check here: https://chromedriver.chromium.org/

**Reply by Yurii397** [reply]
I've found a clue! - https://stackoverflow.com/questions/60313027/setting-download-directory-in-webdriver-using-js
in the new versions of Chrome the prefs should be provided like this 
        prefs: {
          'download': {
            directory_upgrade: true,
            prompt_for_download: false,
            'default_directory': global.downloadFolder,
          }
I'm still wondering what is the difference between headless and headed mode. why it's working differently. Probably will ask there
@christian-bromann thank you for the help!
